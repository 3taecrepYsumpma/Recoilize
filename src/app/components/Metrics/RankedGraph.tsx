import React, {useEffect, useRef} from 'react';
import * as d3 from 'd3';
import {dataDurationArr} from '../../../types';

interface RankedGraphProps {
  data: dataDurationArr;
  width?: number;
  height?: number;
}

const RankedGraph: React.FC<RankedGraphProps> = ({
  data,
  width,
  height,
}: RankedGraphProps) => {
  const svgRef = useRef();
  useEffect(() => {
    document.getElementById('canvas').innerHTML = '';
    // set the dimensions and margins of the graph
    let left = 80;
    if (length > 13) {
      left = 100;
    }
    if (length > 17) {
      left = 120;
    }

    const margin = {top: 20, right: 20, bottom: 30, left};
    // set range for y scale
    const y = d3.scaleBand().range([height, 0]).padding(0.2);
    // set range for x scale
    const x = d3.scaleLinear().range([0, width * 0.8]);
    // set range for durations
    const z = d3.scaleBand().range([height, 0]).padding(0.2);
    // determines the color based on actualDuration
    function colorPicker(data: any) {
      if (data <= 1) return '#51a8f0';
      else if (data <= 2) return '#3a7bb0';
      else return '#2d608a';
    }
    // append the svg object to the body of the page
    // append a 'group' element to 'svg'
    // moves the 'group' element to the top left margin
    const svg = d3
      .select(svgRef.current)
      .classed('svg-container', true)
      .append('svg')
      .attr('class', 'chart')
      .attr('viewBox', '0 0 600 490')
      .attr('preserveAspectRatio', 'xMinYMin meet')
      .classed('svg-content-responsive', true)
      .call(
        d3.zoom().on('zoom', function () {
          svg.attr('transform', d3.event.transform);
        }),
      )
      .append('g')
      .attr('transform', 'translate(' + margin.left + ',' + margin.top + ')');
    // Scale the range of the data in the domain
    x.domain([
      0,
      d3.max(data, (d: any) => {
        return d.actualDuration;
      }),
    ]);
    // Scale the range of the data across the y-axis
    y.domain(
      data.map((d: any, i) => {
        return d.name + '-' + i;
      }),
    );
    // Scale actualDuration with the y-axis
    z.domain(
      data.map((d: any, i) => {
        return d.actualDuration.toFixed(2) + 'ms' + '-' + i;
      }),
    );
    // append the rectangles for the bar chart
    svg
      .selectAll('.bar')
      .data(data)
      .enter()
      .append('rect')
      .attr('class', 'bar')
      .on('mouseover', function () {
        d3.select(this).attr('opacity', '0.85');
        const backgroundConnection = chrome.runtime.connect();
        const barName = this.data;
        const payload = {
          action: 'mouseover',
          tabId: chrome.devtools.inspectedWindow.tabId,
          payload: barName,
        };
        backgroundConnection.postMessage(payload);
      })
      // .transition()
      // .duration(1300)
      // .delay((d: any,i: any) => i * 100)
      .attr('width', function (d: any) {
        return x(d.actualDuration);
      })
      .attr('fill', function (d: any) {
        return colorPicker(d.actualDuration);
      })
      .attr('y', function (d: any, i: any) {
        return y(d.name + '-' + i);
      })
      .attr('height', y.bandwidth());
    // add x axis
    svg
      .append('g')
      .attr('transform', 'translate(0,' + height + ')')
      .call(d3.axisBottom(x));
    // add y axis to able to have duplicate strings
    const yAxis = d3.axisLeft(y).tickFormat(function (d: any) {
      return d.split('-')[0];
    });
    yAxis(svg.append('g'));
    // add z axis to display all duration times
    const zAxis = d3.axisRight(z).tickFormat(function (d: any) {
      return d.split('-')[0];
    });
    zAxis(svg.append('g'));
    // svg.append('g')
    //   .call(d3.axisRight(z));
  }, [data]);

  return (
    <div data-testid="canvas" id="stateGraphContainer">
      <button
        className="save-series-button"
        onClick={e => {
          console.log('save button has been clicked');
        }}>
        Save Series
      </button>
      <svg id="canvas" ref={svgRef}></svg>
    </div>
  );
};

export default RankedGraph;
