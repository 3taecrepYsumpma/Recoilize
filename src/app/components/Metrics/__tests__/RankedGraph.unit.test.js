import React from 'react';
import RankedGraph from '../RankedGraph';
import {render, cleanup, screen} from '@testing-library/react';
import '@testing-library/dom';
import '@testing-library/jest-dom'


//create empty mock data
const mockNoValue = [
  {
    name: '', 
    actualDuration: 0
  }
]

//create mock data with name and actualDuration properties
const mock = [
  {
    name: 'atom1', 
    actualDuration: 0.1345
  },
  {
    name: 'atom2', 
    actualDuration: 0.23456
  },
  {
    name: 'atom3', 
    actualDuration: 1.2357
  },
  
]

//mock heigh and width
const mockHeight = 50
const mockWidth = 30

afterEach(cleanup);

describe('Ranked graph displays correct information', () => {
  it('should render data, height, and width', () => {
    const {asFragment} = render(
      <RankedGraph
        data={mock}
        width={mockWidth}
        height={mockHeight}
      />
    )
    //checking if new component being rendered matches the saved snapshot
    expect(asFragment()).toMatchSnapshot();
  });

  //check that labels are being properly passed in 
  it('should contain data with name property', () => {
    // render ranked graph
    render(
      <RankedGraph
        data={mock}
        width={mockWidth}
        height={mockHeight}
      />
    )
    //ensure that a name property is being passed in and can be found 
    const atom1 = screen.getByText('atom1')
    const atom2 = screen.getByText('atom2')
    const atom3 = screen.getByText('atom3')

    expect(atom1).toBeVisible();
    expect(atom2).toBeVisible();
    expect(atom3).toBeVisible();
  });
  it('should contain data with actualDuration property', () => {
    // render ranked graph
    render(
      <RankedGraph
      data={mock}
      width={mockWidth}
      height={mockHeight}
      />
      )
      //ensure that an actualDuration property is being passed in and can be found
      const atom1Duration = screen.getByText('0.13ms')
      const atom2Duration = screen.getByText('0.23ms')
      const atom3Duration = screen.getByText('1.24ms')
      
      expect(atom1Duration).toBeVisible();
      expect(atom2Duration).toBeVisible();
      expect(atom3Duration).toBeVisible();
      expect(atom3Duration).not.toBe('1.2357ms'); 
  });
});

//This is passing, but it should not be passing
describe('Ranked graph initial rendering', () => {
  it('should not render without initial data', () => {
    const {asFragment} = render(
      <RankedGraph
        data={mockNoValue}
        width={mockWidth}
        height={mockHeight}
      />
    )
    //checking if new component being rendered initially matches snapshot
    expect(asFragment()).toMatchSnapshot();
  });
});