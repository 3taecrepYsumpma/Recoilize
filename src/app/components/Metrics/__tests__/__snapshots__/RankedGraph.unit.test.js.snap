// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ranked graph displays correct information should render data, height, and width 1`] = `
<DocumentFragment>
  <div
    data-testid="canvas"
    id="stateGraphContainer"
  >
    <svg
      class="svg-container"
      id="canvas"
    >
      <svg
        class="chart svg-content-responsive"
        preserveAspectRatio="xMinYMin meet"
        viewBox="-100 0 900 600"
      >
        <g
          transform="translate(80,20)"
        >
          <rect
            class="bar"
            fill="#51a8f0"
            height="12.500000000000002"
            width="2.61228453508133"
            y="34.375"
          />
          <rect
            class="bar"
            fill="#51a8f0"
            height="12.500000000000002"
            width="4.555668851663025"
            y="18.75"
          />
          <rect
            class="bar"
            fill="#3a7bb0"
            height="12.500000000000002"
            width="24"
            y="3.125"
          />
          <g
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="middle"
            transform="translate(0,50)"
          >
            <path
              class="domain"
              d="M0.5,6V0.5H24.5V6"
              stroke="currentColor"
            />
            <g
              class="tick"
              opacity="1"
              transform="translate(0.5,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              >
                0.0
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(2.4422189851905802,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              >
                0.1
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(4.3844379703811605,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              >
                0.2
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(6.326656955571741,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              >
                0.3
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(8.268875940762321,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              >
                0.4
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(10.211094925952901,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              >
                0.5
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(12.153313911143481,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              >
                0.6
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(14.095532896334062,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              >
                0.7
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(16.037751881524642,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              >
                0.8
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(17.97997086671522,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              >
                0.9
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(19.922189851905802,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              >
                1.0
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(21.864408837096384,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              >
                1.1
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(23.806627822286963,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              >
                1.2
              </text>
            </g>
          </g>
          <g
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="end"
          >
            <path
              class="domain"
              d="M-6,50.5H0.5V0.5H-6"
              stroke="currentColor"
            />
            <g
              class="tick"
              opacity="1"
              transform="translate(0,40.625)"
            >
              <line
                stroke="currentColor"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-9"
              >
                atom1
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,25)"
            >
              <line
                stroke="currentColor"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-9"
              >
                atom2
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,9.375)"
            >
              <line
                stroke="currentColor"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-9"
              >
                atom3
              </text>
            </g>
          </g>
          <g
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="start"
          >
            <path
              class="domain"
              d="M6,50.5H0.5V0.5H6"
              stroke="currentColor"
            />
            <g
              class="tick"
              opacity="1"
              transform="translate(0,40.625)"
            >
              <line
                stroke="currentColor"
                x2="6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="9"
              >
                0.13ms
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,25)"
            >
              <line
                stroke="currentColor"
                x2="6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="9"
              >
                0.23ms
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,9.375)"
            >
              <line
                stroke="currentColor"
                x2="6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="9"
              >
                1.24ms
              </text>
            </g>
          </g>
        </g>
      </svg>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Ranked graph initial rendering should not render without initial data 1`] = `
<DocumentFragment>
  <div
    data-testid="canvas"
    id="stateGraphContainer"
  >
    <svg
      class="svg-container"
      id="canvas"
    >
      <svg
        class="chart svg-content-responsive"
        preserveAspectRatio="xMinYMin meet"
        viewBox="-100 0 900 600"
      >
        <g
          transform="translate(80,20)"
        >
          <rect
            class="bar"
            fill="#51a8f0"
            height="33.33333333333333"
            width="12"
            y="8.333333333333336"
          />
          <g
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="middle"
            transform="translate(0,50)"
          >
            <path
              class="domain"
              d="M0.5,6V0.5H24.5V6"
              stroke="currentColor"
            />
            <g
              class="tick"
              opacity="1"
              transform="translate(12.5,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              >
                0
              </text>
            </g>
          </g>
          <g
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="end"
          >
            <path
              class="domain"
              d="M-6,50.5H0.5V0.5H-6"
              stroke="currentColor"
            />
            <g
              class="tick"
              opacity="1"
              transform="translate(0,25)"
            >
              <line
                stroke="currentColor"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-9"
              />
            </g>
          </g>
          <g
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="start"
          >
            <path
              class="domain"
              d="M6,50.5H0.5V0.5H6"
              stroke="currentColor"
            />
            <g
              class="tick"
              opacity="1"
              transform="translate(0,25)"
            >
              <line
                stroke="currentColor"
                x2="6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="9"
              >
                0.00ms
              </text>
            </g>
          </g>
        </g>
      </svg>
    </svg>
  </div>
</DocumentFragment>
`;
